File: backend/app/utils/attendance_processor.py
import os
import logging
import datetime
from pathlib import Path

logger = logging.getLogger("backend_attendance")

def get_date_candidates(s):
    if not s or not isinstance(s, str):
        return []
        
    s = s.strip()
    s_norm = s.replace("/", "-").replace(".", "-").replace(" ", "-")
    
    fmts = [
        "%d-%m-%Y", "%m-%d-%Y", "%Y-%m-%d",
        "%d-%b-%Y", "%b-%d-%Y",
        "%d-%B-%Y", "%B-%d-%Y",
        "%d-%m-%y", "%m-%d-%y", "%y-%m-%d",
        "%d-%b-%y", "%b-%d-%y",
    ]
    
    candidates = []
    for fmt in fmts:
        try:
            dt = datetime.datetime.strptime(s_norm, fmt)
            if 2000 <= dt.year <= 2100:
                d = dt.date()
                if d not in candidates:
                    candidates.append(d)
        except ValueError:
            pass

    for fmt in fmts:
        try:
            dt = datetime.datetime.strptime(s, fmt)
            if 2000 <= dt.year <= 2100:
                d = dt.date()
                if d not in candidates:
                    candidates.append(d)
        except ValueError:
            pass

    return candidates

def resolve_date_sequence(raw_date_strings):
    if not raw_date_strings:
        return []

    candidates_list = []
    for s in raw_date_strings:
        cands = get_date_candidates(s)
        if not cands:
            cands = [datetime.date.today()]
        candidates_list.append(cands)

    n = len(candidates_list)
    valid_sequences = []

    def backtrack(idx, current_seq):
        if len(valid_sequences) >= 1000:
            return
            
        if idx == n:
            valid_sequences.append(list(current_seq))
            return
        
        last_date = current_seq[-1] if current_seq else None
        
        for cand in sorted(candidates_list[idx]):
            if last_date is None or cand >= last_date:
                current_seq.append(cand)
                backtrack(idx + 1, current_seq)
                current_seq.pop()

    backtrack(0, [])
    
    if not valid_sequences:
        return [cands[0] for cands in candidates_list]

    best_seq = valid_sequences[0]
    min_span = None

    for seq in valid_sequences:
        span = (seq[-1] - seq[0]).days
        if min_span is None or span < min_span:
            min_span = span
            best_seq = seq

    return best_seq

def parse_sheets_data(mft_raw, marquee_raw):
    # 1. Parse MFT List
    if not mft_raw or len(mft_raw) < 1:
        raise ValueError("MFT Student list is empty.")
        
    email_key = None
    roll_key = None
    
    for k in mft_raw[0].keys():
        k_low = k.strip().lower()
        if "email" in k_low or "mail" in k_low:
            email_key = k
        elif "rhn" in k_low or "roll" in k_low or "id" in k_low:
            roll_key = k

    if not email_key and mft_raw[0]:
        email_key = list(mft_raw[0].keys())[min(1, len(mft_raw[0])-1)]
    if not roll_key and mft_raw[0]:
        roll_key = list(mft_raw[0].keys())[0]

    mft_students = []
    for row in mft_raw:
        email = str(row.get(email_key, "")).strip()
        roll = str(row.get(roll_key, "")).strip()
        
        enrollment = ""
        if email and "@" in email:
            username = email.split("@")[0]
            if username.isdigit():
                enrollment = username
        
        mft_students.append({
            "raw_email": email,
            "roll_no": roll,
            "extracted_enrollment": enrollment
        })

    # 2. Parse Marquee Sheet
    if not marquee_raw or len(marquee_raw) < 2:
        raise ValueError("Marquee attendance must have at least 2 header rows.")

    row_0 = [str(cell).strip() for cell in marquee_raw[0]]
    row_1 = [str(cell).strip() for cell in marquee_raw[1]]
    data_rows = marquee_raw[2:]

    enroll_col_idx = None
    name_col_idx = None
    email_col_idx = None
    
    for idx, col_name in enumerate(row_0):
        col_name_low = col_name.lower()
        if "enrollment" in col_name_low or "enroll" in col_name_low:
            enroll_col_idx = idx
        elif "name" in col_name_low or "student" in col_name_low:
            name_col_idx = idx
        elif "mail" in col_name_low or "email" in col_name_low:
            email_col_idx = idx

    if enroll_col_idx is None:
        enroll_col_idx = 6
    if name_col_idx is None:
        name_col_idx = 7
    if email_col_idx is None:
        email_col_idx = 14

    raw_headers_list = []
    last_raw_date = ""

    for idx in range(min(len(row_0), len(row_1))):
        if idx < 15:
            continue
            
        cell_0 = row_0[idx].strip()
        cell_1 = row_1[idx].strip()
        
        if cell_0:
            if get_date_candidates(cell_0):
                last_raw_date = cell_0
            else:
                last_raw_date = ""
        
        if last_raw_date:
            raw_headers_list.append((idx, last_raw_date, cell_1))

    raw_date_strings = [item[1] for item in raw_headers_list]
    resolved_dates = resolve_date_sequence(raw_date_strings)

    date_map = {}
    for i, (col_idx, _, session_name) in enumerate(raw_headers_list):
        if i < len(resolved_dates):
            date_obj = resolved_dates[i]
            session_name = session_name if session_name else f"Session {col_idx}"
            if date_obj not in date_map:
                date_map[date_obj] = []
            date_map[date_obj].append((col_idx, session_name))

    marquee_students = []
    for r_idx, row in enumerate(data_rows):
        if len(row) <= max(enroll_col_idx, name_col_idx, email_col_idx):
            continue
            
        enroll = str(row[enroll_col_idx]).strip()
        name = str(row[name_col_idx]).strip()
        email = str(row[email_col_idx]).strip()
        
        if not enroll and not name:
            continue
            
        marquee_students.append({
            "row_idx": r_idx,
            "enrollment": enroll,
            "name": name,
            "email": email,
            "raw_row": row
        })

    return mft_students, marquee_students, date_map

def match_students(mft_students, marquee_students):
    matched_records = []
    mismatched_mft = []

    marquee_by_enroll = {}
    marquee_by_email = {}
    marquee_by_name = {}

    for m_student in marquee_students:
        enroll_clean = m_student["enrollment"].replace(" ", "").strip()
        email_clean = m_student["email"].lower().strip()
        name_clean = m_student["name"].lower().strip()

        if enroll_clean:
            marquee_by_enroll[enroll_clean] = m_student
        if email_clean:
            marquee_by_email[email_clean] = m_student
        if name_clean:
            marquee_by_name[name_clean] = m_student

    for mft in mft_students:
        mft_enroll = mft["extracted_enrollment"].strip()
        mft_roll = mft["roll_no"].strip()
        mft_email = mft["raw_email"].lower().strip()

        matched_marquee = None

        if mft_enroll and mft_enroll in marquee_by_enroll:
            matched_marquee = marquee_by_enroll[mft_enroll]
        elif mft_roll and mft_roll in marquee_by_enroll:
            matched_marquee = marquee_by_enroll[mft_roll]

        if not matched_marquee and mft_email:
            if mft_email in marquee_by_email:
                matched_marquee = marquee_by_email[mft_email]

        if not matched_marquee:
            roll_digits = "".join(filter(str.isdigit, mft_roll))
            if roll_digits and len(roll_digits) >= 4:
                for enroll_key, m_student in marquee_by_enroll.items():
                    if enroll_key.endswith(roll_digits):
                        matched_marquee = m_student
                        break
        
        if matched_marquee:
            matched_records.append({
                "mft_student": mft,
                "marquee_student": matched_marquee
            })
        else:
            mismatched_mft.append(mft)

    return matched_records, mismatched_mft

def compute_attendance(matched_records, mismatched_mft, date_obj, session_cols, mode="Combined (Any)"):
    results = []
    
    if not session_cols:
        raise ValueError(f"No session columns found for date {date_obj}")

    for record in matched_records:
        mft = record["mft_student"]
        marquee = record["marquee_student"]
        raw_row = marquee["raw_row"]

        statuses = []
        for col_idx, sess_name in session_cols:
            if col_idx < len(raw_row):
                val = str(raw_row[col_idx]).strip().lower()
                statuses.append((sess_name, val))
            else:
                statuses.append((sess_name, "absent"))

        final_status = "Absent"
        if mode == "Combined (Any)":
            is_present = any("present" in s[1] for s in statuses)
            final_status = "Present" if is_present else "Absent"
        elif mode == "Combined (All)":
            is_present = all("present" in s[1] for s in statuses)
            final_status = "Present" if is_present else "Absent"
        else:
            val = "absent"
            for sess_name, s_val in statuses:
                if sess_name == mode:
                    val = s_val
                    break
            final_status = "Present" if "present" in val else "Absent"

        results.append({
            "roll_number": mft["roll_no"],
            "enrollment_number": marquee["enrollment"],
            "student_name": marquee["name"],
            "attendance": final_status
        })

    for mft in mismatched_mft:
        results.append({
            "roll_number": mft["roll_no"],
            "enrollment_number": mft["extracted_enrollment"] or "N/A",
            "student_name": "Unknown Student (New/Not Synced)",
            "attendance": "Absent"
        })

    return results

import openpyxl

def load_student_list_excel(file_path):
    if not file_path or not os.path.exists(file_path):
        raise FileNotFoundError(f"Student list Excel file not found or path is empty: {file_path}")
    wb = openpyxl.load_workbook(file_path, data_only=True)
    sheet = wb.active
    rows = list(sheet.rows)
    if len(rows) < 1:
        return []
    headers = [cell.value for cell in rows[0]]
    mft_raw = []
    for row in rows[1:]:
        row_dict = {}
        for col_idx, cell in enumerate(row):
            if col_idx < len(headers):
                header = headers[col_idx]
                if header is not None:
                    row_dict[str(header)] = cell.value
        if row_dict:
            mft_raw.append(row_dict)
    return mft_raw

def load_attendance_logs_excel(file_path):
    if not file_path or not os.path.exists(file_path):
        raise FileNotFoundError(f"Attendance logs Excel file not found or path is empty: {file_path}")
    wb = openpyxl.load_workbook(file_path, data_only=True)
    sheet = wb.active
    marquee_raw = []
    for row in sheet.rows:
        marquee_raw.append([cell.value for cell in row])
    return marquee_raw

File: frontend/src/utils/attendanceProcessor.js
// Robust CSV parser following RFC 4180
export function parseCSV(text) {
  const lines = [];
  let row = [];
  let cell = '';
  let inQuotes = false;

  for (let i = 0; i < text.length; i++) {
    const char = text[i];
    const nextChar = text[i + 1];

    if (inQuotes) {
      if (char === '"') {
        if (nextChar === '"') {
          cell += '"';
          i++; // Skip double quote escape
        } else {
          inQuotes = false;
        }
      } else {
        cell += char;
      }
    } else {
      if (char === '"') {
        inQuotes = true;
      } else if (char === ',') {
        row.push(cell);
        cell = '';
      } else if (char === '\n' || char === '\r') {
        row.push(cell);
        if (row.length > 0 && row.some(c => c !== '')) {
          lines.push(row);
        }
        row = [];
        cell = '';
        if (char === '\r' && nextChar === '\n') {
          i++; // Skip \n after \r
        }
      } else {
        cell += char;
      }
    }
  }
  if (cell || row.length > 0) {
    row.push(cell);
    lines.push(row);
  }
  return lines;
}

// Clean and normalize date string parses
export function interpretDateString(s) {
  if (!s) return [];
  const sNorm = s.trim().replace(/\//g, "-").replace(/\./g, "-").replace(/ /g, "-");
  
  // Format: YYYY-MM-DD
  const ymdMatch = sNorm.match(/^(\d{4})-(\d{1,2})-(\d{1,2})$/);
  if (ymdMatch) {
    const y = parseInt(ymdMatch[1]);
    const m = parseInt(ymdMatch[2]);
    const d = parseInt(ymdMatch[3]);
    if (m >= 1 && m <= 12 && d >= 1 && d <= 31) {
      return [`${y}-${String(m).padStart(2, '0')}-${String(d).padStart(2, '0')}`];
    }
  }

  // Format: DD-MM-YYYY or MM-DD-YYYY
  const dmyMatch = sNorm.match(/^(\d{1,2})-(\d{1,2})-(\d{4})$/);
  if (dmyMatch) {
    const p1 = parseInt(dmyMatch[1]);
    const p2 = parseInt(dmyMatch[2]);
    const y = parseInt(dmyMatch[3]);

    const results = [];
    
    // Case 1: MM-DD-YYYY
    if (p1 >= 1 && p1 <= 12 && p2 >= 1 && p2 <= 31) {
      results.push(`${y}-${String(p1).padStart(2, '0')}-${String(p2).padStart(2, '0')}`);
    }
    
    // Case 2: DD-MM-YYYY
    if (p1 >= 1 && p1 <= 31 && p2 >= 1 && p2 <= 12) {
      const formatted = `${y}-${String(p2).padStart(2, '0')}-${String(p1).padStart(2, '0')}`;
      if (!results.includes(formatted)) {
        results.push(formatted);
      }
    }
    
    return results;
  }

  // Standard JS parser
  try {
    const dt = new Date(sNorm);
    if (!isNaN(dt.getTime())) {
      const y = dt.getFullYear();
      const m = dt.getMonth() + 1;
      const d = dt.getDate();
      if (y >= 2000 && y <= 2100) {
        return [`${y}-${String(m).padStart(2, '0')}-${String(d).padStart(2, '0')}`];
      }
    }
  } catch (e) {}

  return [];
}

// Greedy chronological date sequence resolver
function resolveChronologicalDates(rawHeadersList) {
  let lastDate = null;
  const resolved = [];

  for (const item of rawHeadersList) {
    const candidates = interpretDateString(item.rawDate);
    if (candidates.length === 0) {
      resolved.push(null);
      continue;
    }

    if (lastDate === null) {
      const chosen = candidates[0];
      resolved.push(chosen);
      lastDate = chosen;
    } else {
      const validCandidates = candidates.filter(c => c >= lastDate);
      if (validCandidates.length > 0) {
        let best = validCandidates[0];
        let minDiff = null;
        for (const cand of validCandidates) {
          const diff = new Date(cand) - new Date(lastDate);
          if (minDiff === null || diff < minDiff) {
            minDiff = diff;
            best = cand;
          }
        }
        resolved.push(best);
        lastDate = best;
      } else {
        const chosen = candidates[0];
        resolved.push(chosen);
        lastDate = chosen;
      }
    }
  }
  return resolved;
}

export function parseCSVMetadata(studentCSV, attendanceCSV) {
  // 1. Parse Student List CSV
  const studentRows = parseCSV(studentCSV);
  if (studentRows.length === 0) {
    throw new Error("Student List CSV is empty.");
  }

  const headers = studentRows[0].map(h => String(h).trim());
  let emailKeyIdx = null;
  let rollKeyIdx = null;

  headers.forEach((h, idx) => {
    const hLow = h.toLowerCase();
    if (hLow.includes("email") || hLow.includes("mail")) {
      emailKeyIdx = idx;
    } else if (hLow.includes("rhn") || hLow.includes("roll") || hLow.includes("id")) {
      rollKeyIdx = idx;
    }
  });

  if (emailKeyIdx === null && headers.length > 0) emailKeyIdx = Math.min(1, headers.length - 1);
  if (rollKeyIdx === null && headers.length > 0) rollKeyIdx = 0;

  const mftStudents = studentRows.slice(1).map(row => {
    const email = emailKeyIdx !== null ? String(row[emailKeyIdx] || "").trim() : "";
    const roll = rollKeyIdx !== null ? String(row[rollKeyIdx] || "").trim() : "";
    
    let enrollment = "";
    if (email && email.includes("@")) {
      const uname = email.split("@")[0];
      if (/^\d+$/.test(uname)) {
        enrollment = uname;
      }
    }
    return {
      raw_email: email,
      roll_no: roll,
      extracted_enrollment: enrollment
    };
  }).filter(s => s.raw_email || s.roll_no); // Ignore blank rows

  // 2. Parse Attendance Log CSV
  const marqueeRaw = parseCSV(attendanceCSV);
  if (marqueeRaw.length < 1) {
    throw new Error("Attendance log CSV is empty.");
  }

  let row0 = (marqueeRaw[0] || []).map(cell => String(cell === null || cell === undefined ? "" : cell).trim());
  let row1 = marqueeRaw.length > 1 
    ? (marqueeRaw[1] || []).map(cell => String(cell === null || cell === undefined ? "" : cell).trim()) 
    : [];
  let dataRows = marqueeRaw.slice(2);

  // Auto-detect Student Info columns
  let enrollColIdx = null;
  let nameColIdx = null;
  let emailColIdx = null;

  for (let i = 0; i < row0.length; i++) {
    const colNameLow = row0[i].toLowerCase();
    if (colNameLow.includes("enrollment") || colNameLow.includes("enroll")) {
      enrollColIdx = i;
    } else if (colNameLow.includes("name") || colNameLow.includes("student")) {
      nameColIdx = i;
    } else if (colNameLow.includes("mail") || colNameLow.includes("email")) {
      emailColIdx = i;
    }
  }

  if (enrollColIdx === null) enrollColIdx = 6;
  if (nameColIdx === null) nameColIdx = 7;
  if (emailColIdx === null) emailColIdx = 14;

  // Auto-detect transactional vs. cross-tab date columns
  let transactionalDateColIdx = -1;
  for (let i = 0; i < row0.length; i++) {
    const colNameLow = row0[i].toLowerCase();
    if ((colNameLow === "date" || colNameLow === "attendance date" || colNameLow === "class date" || colNameLow === "attendance_date" || colNameLow === "class_date") && !colNameLow.includes("birth") && !colNameLow.includes("dob")) {
      transactionalDateColIdx = i;
      break;
    }
  }

  // Pivot transactional structure to cross-tab if detected
  if (transactionalDateColIdx !== -1) {
    let transactionalSessionColIdx = -1;
    let transactionalStatusColIdx = -1;
    for (let i = 0; i < row0.length; i++) {
      const colNameLow = row0[i].toLowerCase();
      if (colNameLow.includes("session") || colNameLow.includes("column") || colNameLow.includes("period")) {
        transactionalSessionColIdx = i;
      } else if (colNameLow.includes("status") || colNameLow.includes("attendance") || colNameLow.includes("present") || colNameLow.includes("mark")) {
        transactionalStatusColIdx = i;
      }
    }

    const studentsMap = {};
    const uniqueDates = new Set();
    const uniqueSessions = new Set();

    const rawRows = marqueeRaw.slice(1);
    rawRows.forEach(row => {
      if (row.length <= Math.max(enrollColIdx, nameColIdx, emailColIdx)) return;
      const enroll = String(row[enrollColIdx] || "").trim();
      const name = String(row[nameColIdx] || "").trim();
      const email = String(row[emailColIdx] || "").trim();
      const rawDate = String(row[transactionalDateColIdx] || "").trim();
      
      const resolvedDateCands = interpretDateString(rawDate);
      if (resolvedDateCands.length === 0) return;
      const dateVal = resolvedDateCands[0];

      const sessionVal = transactionalSessionColIdx !== -1 && row[transactionalSessionColIdx]
        ? String(row[transactionalSessionColIdx]).trim()
        : "Session 01";

      const statusVal = transactionalStatusColIdx !== -1 && row[transactionalStatusColIdx]
        ? String(row[transactionalStatusColIdx]).trim()
        : "Present";

      if (!enroll && !name) return;

      const studentKey = enroll || email || name;
      if (!studentsMap[studentKey]) {
        studentsMap[studentKey] = {
          enrollment: enroll,
          name,
          email,
          attendance: {}
        };
      }
      studentsMap[studentKey].attendance[`${dateVal}|${sessionVal}`] = statusVal;
      uniqueDates.add(dateVal);
      uniqueSessions.add(sessionVal);
    });

    const datesList = Array.from(uniqueDates).sort();
    const sessionsList = Array.from(uniqueSessions).sort();

    const pivotedRow0 = [...row0.slice(0, 15)];
    const pivotedRow1 = [...row1.slice(0, 15)];
    while (pivotedRow0.length < 15) pivotedRow0.push("");
    while (pivotedRow1.length < 15) pivotedRow1.push("");

    const colMapping = [];
    let colIdx = 15;
    datesList.forEach(d => {
      sessionsList.forEach(s => {
        pivotedRow0.push(d);
        pivotedRow1.push(s);
        colMapping.push({ date: d, session: s, colIdx });
        colIdx++;
      });
    });

    const pivotedDataRows = Object.values(studentsMap).map(student => {
      const row = new Array(colIdx).fill("");
      row[enrollColIdx] = student.enrollment;
      row[nameColIdx] = student.name;
      row[emailColIdx] = student.email;
      colMapping.forEach(mapItem => {
        row[mapItem.colIdx] = student.attendance[`${mapItem.date}|${mapItem.session}`] || "Absent";
      });
      return row;
    });

    row0 = pivotedRow0;
    row1 = pivotedRow1;
    dataRows = pivotedDataRows;
  }

  // Cross-tab parsing
  const rawHeadersList = [];
  let lastRawDate = "";

  for (let idx = 0; idx < Math.min(row0.length, row1.length); idx++) {
    if (idx < 15) continue;
    const cell0 = row0[idx].trim();
    const cell1 = row1[idx].trim();

    if (cell0) {
      const cands = interpretDateString(cell0);
      if (cands.length > 0) {
        lastRawDate = cell0;
      } else {
        lastRawDate = "";
      }
    }

    if (lastRawDate) {
      rawHeadersList.push({ colIdx: idx, rawDate: lastRawDate, sessionName: cell1 });
    }
  }

  const resolvedDates = resolveChronologicalDates(rawHeadersList);

  const dateMap = {};
  rawHeadersList.forEach((item, index) => {
    if (index < resolvedDates.length) {
      const dateObj = resolvedDates[index];
      if (dateObj) {
        const sessionName = item.sessionName ? item.sessionName : `Session ${item.colIdx}`;
        if (!dateMap[dateObj]) {
          dateMap[dateObj] = [];
        }
        dateMap[dateObj].push({ colIdx: item.colIdx, sessionName });
      }
    }
  });

  const marqueeStudents = [];
  dataRows.forEach((row, rIdx) => {
    if (row.length <= Math.max(enrollColIdx, nameColIdx, emailColIdx)) return;
    
    const enroll = String(row[enrollColIdx] === null || row[enrollColIdx] === undefined ? "" : row[enrollColIdx]).trim();
    const name = String(row[nameColIdx] === null || row[nameColIdx] === undefined ? "" : row[nameColIdx]).trim();
    const email = String(row[emailColIdx] === null || row[emailColIdx] === undefined ? "" : row[emailColIdx]).trim();

    if (!enroll && !name) return; // Ignore blank rows

    marqueeStudents.push({
      row_idx: rIdx,
      enrollment: enroll,
      name,
      email,
      raw_row: row
    });
  });

  const dates = Object.keys(dateMap).sort((a, b) => b.localeCompare(a));
  const dateSessions = {};
  dates.forEach(d => {
    const sessions = dateMap[d].map(item => item.sessionName);
    dateSessions[d] = ["Combined (Any)", "Combined (All)"].concat(sessions);
  });

  // Print required debugging logs
  console.log("=== CSV Parsing Debug Logs ===");
  console.log("CSV downloaded successfully");
  console.log("Number of rows loaded in Student List:", studentRows.length);
  console.log("Number of rows loaded in Attendance Logs:", marqueeRaw.length);
  console.log("Detected column names (Student):", headers);
  console.log("Detected column names (Attendance):", row0.slice(0, 15));
  
  if (transactionalDateColIdx !== -1) {
    console.log("Detected date column: Transactional column named '" + row0[transactionalDateColIdx] + "' at index " + transactionalDateColIdx);
  } else {
    console.log("Detected date column: Cross-tab date columns starting from index 15");
  }
  
  console.log("Available dates:", dates);
  console.log("Available session columns:", dateSessions);
  console.log("===============================");

  return {
    dates,
    dateSessions,
    mftStudents,
    marqueeStudents,
    dateMap
  };
}

export function compileAttendancePreview(mftStudents, marqueeStudents, dateMap, selectedDate, selectedSession) {
  const marqueeByEnroll = {};
  const marqueeByEmail = {};
  const marqueeByName = {};

  marqueeStudents.forEach(m => {
    const enrollClean = m.enrollment.replace(/ /g, "").trim();
    const emailClean = m.email.toLowerCase().trim();
    const nameClean = m.name.toLowerCase().trim();

    if (enrollClean) marqueeByEnroll[enrollClean] = m;
    if (emailClean) marqueeByEmail[emailClean] = m;
    if (nameClean) marqueeByName[nameClean] = m;
  });

  const matchedRecords = [];
  const mismatchedMft = [];

  mftStudents.forEach(mft => {
    const mftEnroll = mft.extracted_enrollment.trim();
    const mftRoll = mft.roll_no.trim();
    const mftEmail = mft.raw_email.toLowerCase().trim();

    let matchedMarquee = null;

    if (mftEnroll && marqueeByEnroll[mftEnroll]) {
      matchedMarquee = marqueeByEnroll[mftEnroll];
    } else if (mftRoll && marqueeByEnroll[mftRoll]) {
      matchedMarquee = marqueeByEnroll[mftRoll];
    }

    if (!matchedMarquee && mftEmail) {
      if (marqueeByEmail[mftEmail]) {
        matchedMarquee = marqueeByEmail[mftEmail];
      }
    }

    if (!matchedMarquee) {
      const rollDigits = mftRoll.replace(/\D/g, "");
      if (rollDigits && rollDigits.length >= 4) {
        const keys = Object.keys(marqueeByEnroll);
        for (const enrollKey of keys) {
          if (enrollKey.endsWith(rollDigits)) {
            matchedMarquee = marqueeByEnroll[enrollKey];
            break;
          }
        }
      }
    }

    if (matchedMarquee) {
      matchedRecords.push({ mft_student: mft, marquee_student: matchedMarquee });
    } else {
      mismatchedMft.push(mft);
    }
  });

  const allSessions = dateMap[selectedDate] || [];
  let sessionCols = [];

  if (selectedSession === "Combined (Any)" || selectedSession === "Combined (All)") {
    sessionCols = allSessions;
  } else {
    sessionCols = allSessions.filter(s => s.sessionName === selectedSession);
  }

  const records = [];

  matchedRecords.forEach(record => {
    const mft = record.mft_student;
    const marquee = record.marquee_student;
    const rawRow = marquee.raw_row;

    const statuses = sessionCols.map(col => {
      if (col.colIdx < rawRow.length) {
        const val = String(rawRow[col.colIdx] === null || rawRow[col.colIdx] === undefined ? "" : rawRow[col.colIdx]).trim().toLowerCase();
        return { session: col.sessionName, status: val };
      }
      return { session: col.sessionName, status: "absent" };
    });

    let finalStatus = "Absent";
    if (selectedSession === "Combined (Any)") {
      const isPresent = statuses.some(s => s.status.includes("present"));
      finalStatus = isPresent ? "Present" : "Absent";
    } else if (selectedSession === "Combined (All)") {
      const isPresent = statuses.every(s => s.status.includes("present"));
      finalStatus = isPresent ? "Present" : "Absent";
    } else {
      const targetVal = statuses[0]?.status || "absent";
      finalStatus = targetVal.includes("present") ? "Present" : "Absent";
    }

    records.push({
      roll_number: mft.roll_no,
      enrollment_number: marquee.enrollment,
      student_name: marquee.name,
      attendance: finalStatus
    });
  });

  mismatchedMft.forEach(mft => {
    records.push({
      roll_number: mft.roll_no,
      enrollment_number: mft.extracted_enrollment || "N/A",
      student_name: "Unknown Student (New/Not Synced)",
      attendance: "Absent"
    });
  });

  const total = records.length;
  const present = records.filter(r => r.attendance === "Present").length;
  const absent = total - present;
  const rate = total > 0 ? parseFloat(((present / total) * 100).toFixed(1)) : 0;

  return {
    records,
    summary: {
      total,
      present,
      absent,
      rate
    }
  };
}
